RevenueCatUI/Plugins/iOS/RevenueCatUI.m

```diff
@@ -1,31 +1,170 @@
+static NSString *const kRCUIOptionPresentedOfferingContext = @"presentedOfferingContext";
+static NSString *const kRCUIOptionDisplayCloseButton = @"displayCloseButton";
+static NSString *const kRCUIPresentedOfferingIdentifierKey = @"offeringIdentifier";
 
-void rcui_presentPaywall(const char* offeringIdentifier, bool displayCloseButton, RCUIPaywallResultCallback callback) {
-    NSLog(@"[RevenueCatUI][iOS] presentPaywall(offering=%@, closeButton=%@)",
-          offeringIdentifier ? [NSString stringWithUTF8String:offeringIdentifier] : @"<null>",
-          displayCloseButton ? @"true" : @"false");
-    if (callback) {
-        callback("CANCELLED|Stub: no native UI");
+static NSString *RCUIStringFromCString(const char *string) {
+    if (string == NULL) {
+        return nil;
     }
+    return [NSString stringWithUTF8String:string];
 }
 
-void rcui_presentPaywallIfNeeded(const char* requiredEntitlementIdentifier, const char* offeringIdentifier, bool displayCloseButton, RCUIPaywallResultCallback callback) {
-    NSLog(@"[RevenueCatUI][iOS] presentPaywallIfNeeded(entitlement=%@, offering=%@, closeButton=%@)",
-          requiredEntitlementIdentifier ? [NSString stringWithUTF8String:requiredEntitlementIdentifier] : @"<null>",
-          offeringIdentifier ? [NSString stringWithUTF8String:offeringIdentifier] : @"<null>",
-          displayCloseButton ? @"true" : @"false");
-    if (callback) {
-        callback("NOT_PRESENTED|Stub: no native UI");
+static NSString *RCUINormalizedResultToken(NSString *resultName) {
+    if (resultName.length == 0) {
+        return @"ERROR";
     }
+
+    NSString *normalized = [[[resultName stringByReplacingOccurrencesOfString:@"_" withString:@""]
+                                       stringByReplacingOccurrencesOfString:@"-" withString:@""]
+                                      lowercaseString];
+
+    if ([normalized isEqualToString:@"purchased"]) {
+        return @"PURCHASED";
+    }
+    if ([normalized isEqualToString:@"restored"]) {
+        return @"RESTORED";
+    }
+    if ([normalized isEqualToString:@"cancelled"] || [normalized isEqualToString:@"usercancelled"]) {
+        return @"CANCELLED";
+    }
+    if ([normalized isEqualToString:@"notpresented"] || [normalized isEqualToString:@"notshown"]) {
+        return @"NOT_PRESENTED";
+    }
+    if ([normalized isEqualToString:@"error"] || [normalized isEqualToString:@"failed"]) {
+        return @"ERROR";
+    }
+
+    return [resultName uppercaseString];
+}
+
+static BOOL RCUIRuntimeSupportsPaywalls(void) {
+    if (@available(iOS 15.0, *)) {
+        return YES;
+    }
+    return NO;
+}
+
+static void RCUIInvokeCallback(RCUIPaywallResultCallback callback, NSString *token, NSString *message) {
+    if (callback == NULL) {
+        return;
+    }
+
+    NSString *payload = token ?: @"ERROR";
+    if (message.length > 0) {
+        payload = [payload stringByAppendingFormat:@"|%@", message];
```

Comment:
looking at `IOSPaywallPresenter` I believe the message is always ignored and only the token is forwarded. Fwiw, in Android I am just sending the "token" as well.